File: apps/ade-api/src/migrate.ts
import process from "node:process";
import { readMigrationConfig } from "./config";
import { runMigrations } from "./db/migrate-runner";
import { createConsoleLogger, runMain } from "./process";

type MainDependencies = {
  env?: NodeJS.ProcessEnv;
  run?: typeof runMigrations;
};

async function main(
  logger = createConsoleLogger(),
  dependencies: MainDependencies = {},
) {
  const config = readMigrationConfig(dependencies.env ?? process.env);
  const run = dependencies.run ?? runMigrations;

  const result = await run({
    connectionString: config.sqlConnectionString,
  });

  for (const migrationName of result.applied) {
    logger.info(`Applied migration: ${migrationName}`);
  }

  for (const migrationName of result.skipped) {
    logger.info(`Skipped migration: ${migrationName}`);
  }

  logger.info(
    `Migration complete. Applied ${String(result.applied.length)}, skipped ${String(result.skipped.length)}.`,
  );
}

if (require.main === module) {
  void runMain(async () => {
    await main();
  });
}

export { main };

File: apps/ade-api/src/bin/migrate.rs
use std::process;

use ade_api::{
    config::{AppConfig, EnvBag},
    db::run_migrations,
    error::AppError,
    init_tracing,
};

#[tokio::main]
async fn main() {
    init_tracing();

    let result: Result<(), AppError> = async {
        let env: EnvBag = std::env::vars().collect();
        let config = AppConfig::from_env(&env)?;
        let applied = run_migrations(&config.sql_connection_string).await?;

        if applied.is_empty() {
            tracing::info!("No SQL migrations were pending.");
        } else {
            tracing::info!(count = applied.len(), "Applied SQL migrations.");
        }

        Ok(())
    }
    .await;

    if let Err(error) = result {
        tracing::error!(error = %error, "ADE SQL migrations failed.");
        process::exit(1);
    }
}
